File: hera/services/economy.ts
/**
 * Account and currency operations: wallet, bank, daily/work payouts.
 *
 * Every mutation writes a row to ``transactions`` so the history command has a
 * source of truth and balances remain auditable.
 */
import { config } from '../config';
import type { Database } from '../database';
import { BankFull, CooldownActive, InsufficientFunds } from '../errors';

export interface Account {
  userId: string;
  guildId: string;
  wallet: number;
  bank: number;
  bankCapacity: number;
  bankLevel: number;
  dailyStreak: number;
  lastDaily: number | null;
  lastWork: number | null;
  lastRob: number | null;
}

export function accountNetWorth(account: Account): number {
  return account.wallet + account.bank;
}

interface AccountRow {
  user_id: number;
  guild_id: number;
  wallet: number;
  bank: number;
  bank_capacity: number;
  bank_level: number;
  daily_streak: number;
  last_daily: number | null;
  last_work: number | null;
  last_rob: number | null;
}

export interface TransactionRow {
  kind: string;
  amount: number;
  balance_after: number;
  note: string | null;
  created_at: number;
}

export interface WealthRow {
  user_id: number;
  total: number;
}

export class EconomyService {
  constructor(private readonly db: Database) {}

  // ------------------------------------------------------------------ reads

  getAccount(userId: string, guildId: string): Account {
    let row = this.db.fetchone<AccountRow>(
      'SELECT * FROM accounts WHERE user_id = ? AND guild_id = ?',
      [userId, guildId],
    );
    if (!row) {
      this.db.transaction(() => {
        this.db.execute(
          `INSERT INTO accounts (user_id, guild_id, wallet, bank, bank_capacity)
           VALUES (?, ?, ?, 0, ?)
           ON CONFLICT (user_id, guild_id) DO NOTHING`,
          [userId, guildId, config.economy.startingWallet, config.economy.startingBankCapacity],
        );
        this.journal(
          userId,
          guildId,
          'starting_balance',
          config.economy.startingWallet,
          config.economy.startingWallet,
          'Welcome grant',
        );
      });
      row = this.db.fetchone<AccountRow>(
        'SELECT * FROM accounts WHERE user_id = ? AND guild_id = ?',
        [userId, guildId],
      );
    }
    return this.rowToAccount(row!);
  }

  history(userId: string, guildId: string, limit = 10): TransactionRow[] {
    return this.db.fetchall<TransactionRow>(
      `SELECT kind, amount, balance_after, note, created_at
       FROM transactions WHERE user_id = ? AND guild_id = ?
       ORDER BY id DESC LIMIT ?`,
      [userId, guildId, limit],
    );
  }

  /** Rank members by total liquid wealth. */
  leaderboard(guildId: string, limit = 10): WealthRow[] {
    return this.db.fetchall<WealthRow>(
      `SELECT user_id, wallet + bank AS total FROM accounts
       WHERE guild_id = ? ORDER BY total DESC LIMIT ?`,
      [guildId, limit],
    );
  }

  // -------------------------------------------------------------- mutations

  /** Add credits. Overflow past bank capacity stays in the wallet. */
  credit(
    userId: string,
    guildId: string,
    amount: number,
    options: { kind: string; note?: string | null; toBank?: boolean },
  ): Account {
    if (amount <= 0) throw new Error('credit amount must be positive');
    this.getAccount(userId, guildId);
    this.db.transaction(() => {
      if (options.toBank) {
        this.db.execute(
          `UPDATE accounts
           SET bank = MIN(bank + ?, bank_capacity),
               wallet = wallet + MAX(0, ? - (bank_capacity - bank)),
               updated_at = unixepoch('subsec')
           WHERE user_id = ? AND guild_id = ?`,
          [amount, amount, userId, guildId],
        );
      } else {
        this.db.execute(
          `UPDATE accounts SET wallet = wallet + ?, updated_at = unixepoch('subsec')
           WHERE user_id = ? AND guild_id = ?`,
          [amount, userId, guildId],
        );
      }
      const balance = this.totalLocked(userId, guildId);
      this.journal(userId, guildId, options.kind, amount, balance, options.note ?? null);
    });
    return this.getAccount(userId, guildId);
  }

  /** Remove credits, raising :class:`InsufficientFunds` when short. */
  debit(
    userId: string,
    guildId: string,
    amount: number,
    options: { kind: string; note?: string | null; fromBank?: boolean },
  ): Account {
    if (amount <= 0) throw new Error('debit amount must be positive');
    const account = this.getAccount(userId, guildId);
    const available = options.fromBank ? account.bank : account.wallet;
    if (available < amount) throw new InsufficientFunds(amount, available);
    this.db.transaction(() => {
      const column = options.fromBank ? 'bank' : 'wallet';
      this.db.execute(
        `UPDATE accounts SET ${column} = ${column} - ?, updated_at = unixepoch('subsec')
         WHERE user_id = ? AND guild_id = ?`,
        [amount, userId, guildId],
      );
      const balance = this.totalLocked(userId, guildId);
      this.journal(userId, guildId, options.kind, -amount, balance, options.note ?? null);
    });
    return this.getAccount(userId, guildId);
  }

  /** Move credits between wallet and bank in a single transaction. */
  transfer(userId: string, guildId: string, amount: number): Account {
    const account = this.getAccount(userId, guildId);
    if (amount <= 0) throw new Error('transfer amount must be positive');
    if (account.wallet < amount) throw new InsufficientFunds(amount, account.wallet);
    if (account.bank + amount > account.bankCapacity) throw new BankFull(account.bankCapacity);
    this.db.transaction(() => {
      this.db.execute(
        `UPDATE accounts SET wallet = wallet - ?, bank = bank + ?, updated_at = unixepoch('subsec')
         WHERE user_id = ? AND guild_id = ?`,
        [amount, amount, userId, guildId],
      );
      this.journal(userId, guildId, 'deposit', 0, accountNetWorth(account), `Deposited ${amount.toLocaleString()}`);
    });
    return this.getAccount(userId, guildId);
  }

  withdraw(userId: string, guildId: string, amount: number): Account {
    const account = this.getAccount(userId, guildId);
    if (amount <= 0) throw new Error('withdrawal amount must be positive');
    if (account.bank < amount) throw new InsufficientFunds(amount, account.bank);
    this.db.transaction(() => {
      this.db.execute(
        `UPDATE accounts SET wallet = wallet + ?, bank = bank - ?, updated_at = unixepoch('subsec')
         WHERE user_id = ? AND guild_id = ?`,
        [amount, amount, userId, guildId],
      );
      this.journal(userId, guildId, 'withdraw', 0, accountNetWorth(account), `Withdrew ${amount.toLocaleString()}`);
    });
    return this.getAccount(userId, guildId);
  }

  /** Purchase the next bank tier. Returns the account and the cost paid. */
  upgradeBank(userId: string, guildId: string): [Account, number] {
    const account = this.getAccount(userId, guildId);
    const cost = config.economy.bankUpgradeBaseCost * account.bankLevel;
    if (account.wallet < cost) throw new InsufficientFunds(cost, account.wallet);
    this.db.transaction(() => {
      this.db.execute(
        `UPDATE accounts
         SET wallet = wallet - ?, bank_level = bank_level + 1,
             bank_capacity = bank_capacity + ?, updated_at = unixepoch('subsec')
         WHERE user_id = ? AND guild_id = ?`,
        [cost, config.economy.bankUpgradeCapacity, userId, guildId],
      );
      const balance = this.totalLocked(userId, guildId);
      this.journal(
        userId,
        guildId,
        'bank_upgrade',
        -cost,
        balance,
        `Bank upgraded to level ${account.bankLevel + 1}`,
      );
    });
    return [this.getAccount(userId, guildId), cost];
  }

  // ------------------------------------------------------------- cooldowns

  private static remaining(last: number | null, cooldown: number): number {
    if (last === null) return 0.0;
    return Math.max(0.0, last + cooldown - Date.now() / 1000);
  }

  work(userId: string, guildId: string): [number, Account] {
    const account = this.getAccount(userId, guildId);
    const remaining = EconomyService.remaining(account.lastWork, config.economy.workCooldownSeconds);
    if (remaining > 0) throw new CooldownActive(remaining);
    const payout = this.randomInt(config.economy.workMin, config.economy.workMax);
    this.db.execute('UPDATE accounts SET last_work = ? WHERE user_id = ? AND guild_id = ?', [
      Date.now() / 1000,
      userId,
      guildId,
    ]);
    const updated = this.credit(userId, guildId, payout, { kind: 'work', note: 'Shift wages' });
    return [payout, updated];
  }

  /** Claim the daily reward. Returns [payout, streak, account]. */
  daily(userId: string, guildId: string): [number, number, Account] {
    const account = this.getAccount(userId, guildId);
    const remaining = EconomyService.remaining(account.lastDaily, config.economy.dailyCooldownSeconds);
    if (remaining > 0) throw new CooldownActive(remaining);

    // The streak survives one missed day; beyond that it resets.
    const now = Date.now() / 1000;
    let streak: number;
    if (
      account.lastDaily === null ||
      now - account.lastDaily > config.economy.dailyCooldownSeconds * 2
    ) {
      streak = 1;
    } else {
      streak = Math.min(account.dailyStreak + 1, config.economy.dailyStreakCap);
    }

    const payout = config.economy.dailyAmount + config.economy.dailyStreakBonus * (streak - 1);
    this.db.transaction(() => {
      this.db.execute(
        `UPDATE accounts SET last_daily = ?, daily_streak = ?, updated_at = unixepoch('subsec')
         WHERE user_id = ? AND guild_id = ?`,
        [now, streak, userId, guildId],
      );
      this.db.execute(
        `UPDATE accounts SET bank = MIN(bank + ?, bank_capacity)
         WHERE user_id = ? AND guild_id = ?`,
        [payout, userId, guildId],
      );
      const balance = this.totalLocked(userId, guildId);
      this.journal(userId, guildId, 'daily', payout, balance, `Daily reward (streak ${streak})`);
    });
    return [payout, streak, this.getAccount(userId, guildId)];
  }

  /** Attempt a robbery. Returns [succeeded, amount, robber account]. */
  rob(userId: string, guildId: string, targetId: string): [boolean, number, Account] {
    if (userId === targetId) throw new Error('you cannot rob yourself');
    const robber = this.getAccount(userId, guildId);
    const remaining = EconomyService.remaining(robber.lastRob, config.economy.robCooldownSeconds);
    if (remaining > 0) throw new CooldownActive(remaining);
    const victim = this.getAccount(targetId, guildId);

    this.db.execute('UPDATE accounts SET last_rob = ? WHERE user_id = ? AND guild_id = ?', [
      Date.now() / 1000,
      userId,
      guildId,
    ]);

    if (victim.wallet < 100 || Math.random() > config.economy.robSuccessChance) {
      const fine = Math.min(config.economy.robFine, robber.wallet);
      if (fine > 0) {
        this.debit(userId, guildId, fine, { kind: 'rob_fine', note: 'Failed robbery' });
      }
      return [false, fine, this.getAccount(userId, guildId)];
    }

    let stolen = Math.trunc(
      victim.wallet * (0.05 + Math.random() * (config.economy.robMaxStealFraction - 0.05)),
    );
    stolen = Math.max(1, Math.min(stolen, victim.wallet));
    this.db.transaction(() => {
      this.db.execute('UPDATE accounts SET wallet = wallet - ? WHERE user_id = ? AND guild_id = ?', [
        stolen,
        targetId,
        guildId,
      ]);
      this.db.execute('UPDATE accounts SET wallet = wallet + ? WHERE user_id = ? AND guild_id = ?', [
        stolen,
        userId,
        guildId,
      ]);
      const victimBalance = this.totalLocked(targetId, guildId);
      this.journal(targetId, guildId, 'robbed', -stolen, victimBalance, 'Robbery');
      const robberBalance = this.totalLocked(userId, guildId);
      this.journal(userId, guildId, 'robbery', stolen, robberBalance, 'Successful robbery');
    });
    return [true, stolen, this.getAccount(userId, guildId)];
  }

  // ------------------------------------------------------------- internals

  private totalLocked(userId: string, guildId: string): number {
    const row = this.db.fetchone<Record<string, number>>(
      'SELECT wallet + bank AS total FROM accounts WHERE user_id = ? AND guild_id = ?',
      [userId, guildId],
    );
    return row ? Number(row.total) : 0;
  }

  private journal(
    userId: string,
    guildId: string,
    kind: string,
    amount: number,
    balanceAfter: number,
    note: string | null,
  ): void {
    this.db.execute(
      `INSERT INTO transactions (user_id, guild_id, kind, amount, balance_after, note)
       VALUES (?, ?, ?, ?, ?, ?)`,
      [userId, guildId, kind, amount, balanceAfter, note],
    );
  }

  private randomInt(min: number, max: number): number {
    return min + Math.floor(Math.random() * (max - min + 1));
  }

  private rowToAccount(row: AccountRow): Account {
    return {
      userId: String(row.user_id),
      guildId: String(row.guild_id),
      wallet: Number(row.wallet),
      bank: Number(row.bank),
      bankCapacity: Number(row.bank_capacity),
      bankLevel: Number(row.bank_level),
      dailyStreak: Number(row.daily_streak),
      lastDaily: row.last_daily === null ? null : Number(row.last_daily),
      lastWork: row.last_work === null ? null : Number(row.last_work),
      lastRob: row.last_rob === null ? null : Number(row.last_rob),
    };
  }
}

File: hera/services/economy.py
"""Account and currency operations: wallet, bank, daily/work payouts.

Balances live in :mod:`hera.services.stocks`'s database but conceptually belong
to the wider economy. Every mutation writes a row to ``transactions`` so the
history command has a source of truth and balances remain auditable.
"""

from __future__ import annotations

import random
import time
from dataclasses import dataclass

from ..config import config
from ..database import Database
from ..errors import BankFull, CooldownActive, InsufficientFunds


@dataclass
class Account:
    user_id: int
    guild_id: int
    wallet: int
    bank: int
    bank_capacity: int
    bank_level: int
    daily_streak: int
    last_daily: float | None
    last_work: float | None
    last_rob: float | None

    @property
    def net_worth(self) -> int:
        return self.wallet + self.bank

    @property
    def total(self) -> int:
        return self.wallet + self.bank


class EconomyService:
    """Owns the ``accounts`` and ``transactions`` tables."""

    def __init__(self, db: Database) -> None:
        self.db = db

    # ------------------------------------------------------------------ reads

    async def get_account(self, user_id: int, guild_id: int) -> Account:
        row = await self.db.fetchone(
            "SELECT * FROM accounts WHERE user_id = ? AND guild_id = ?",
            (user_id, guild_id),
        )
        if row is None:
            async with self.db.transaction() as conn:
                await conn.execute(
                    """
                    INSERT INTO accounts (user_id, guild_id, wallet, bank, bank_capacity)
                    VALUES (?, ?, ?, 0, ?)
                    ON CONFLICT (user_id, guild_id) DO NOTHING
                    """,
                    (
                        user_id,
                        guild_id,
                        config.economy.starting_wallet,
                        config.economy.starting_bank_capacity,
                    ),
                )
                await conn.execute(
                    """
                    INSERT INTO transactions
                        (user_id, guild_id, kind, amount, balance_after, note)
                    VALUES (?, ?, 'starting_balance', ?, ?, 'Welcome grant')
                    """,
                    (user_id, guild_id, config.economy.starting_wallet, config.economy.starting_wallet),
                )
            row = await self.db.fetchone(
                "SELECT * FROM accounts WHERE user_id = ? AND guild_id = ?",
                (user_id, guild_id),
            )
        return self._row_to_account(row)

    async def history(self, user_id: int, guild_id: int, limit: int = 10):
        return await self.db.fetchall(
            """
            SELECT kind, amount, balance_after, note, created_at
            FROM transactions
            WHERE user_id = ? AND guild_id = ?
            ORDER BY id DESC
            LIMIT ?
            """,
            (user_id, guild_id, limit),
        )

    async def leaderboard(self, guild_id: int, limit: int = 10):
        """Rank members by total liquid wealth."""
        return await self.db.fetchall(
            """
            SELECT user_id, wallet + bank AS total
            FROM accounts
            WHERE guild_id = ?
            ORDER BY total DESC
            LIMIT ?
            """,
            (guild_id, limit),
        )

    # -------------------------------------------------------------- mutations

    async def credit(
        self,
        user_id: int,
        guild_id: int,
        amount: int,
        *,
        kind: str,
        note: str | None = None,
        to_bank: bool = False,
    ) -> Account:
        """Add credits. Overflow past bank capacity stays in the wallet."""
        if amount <= 0:
            raise ValueError("credit amount must be positive")
        await self.get_account(user_id, guild_id)
        async with self.db.transaction() as conn:
            if to_bank:
                await conn.execute(
                    """
                    UPDATE accounts
                    SET bank = MIN(bank + ?, bank_capacity),
                        wallet = wallet + MAX(0, ? - (bank_capacity - bank)),
                        updated_at = unixepoch('subsec')
                    WHERE user_id = ? AND guild_id = ?
                    """,
                    (amount, amount, user_id, guild_id),
                )
            else:
                await conn.execute(
                    """
                    UPDATE accounts
                    SET wallet = wallet + ?, updated_at = unixepoch('subsec')
                    WHERE user_id = ? AND guild_id = ?
                    """,
                    (amount, user_id, guild_id),
                )
            balance = await self._total_locked(conn, user_id, guild_id)
            await self._journal(conn, user_id, guild_id, kind, amount, balance, note)
        return await self.get_account(user_id, guild_id)

    async def debit(
        self,
        user_id: int,
        guild_id: int,
        amount: int,
        *,
        kind: str,
        note: str | None = None,
        from_bank: bool = False,
    ) -> Account:
        """Remove credits, raising :class:`InsufficientFunds` when short."""
        if amount <= 0:
            raise ValueError("debit amount must be positive")
        account = await self.get_account(user_id, guild_id)
        available = account.bank if from_bank else account.wallet
        if available < amount:
            raise InsufficientFunds(amount, available)
        async with self.db.transaction() as conn:
            column = "bank" if from_bank else "wallet"
            await conn.execute(
                f"""
                UPDATE accounts
                SET {column} = {column} - ?, updated_at = unixepoch('subsec')
                WHERE user_id = ? AND guild_id = ?
                """,
                (amount, user_id, guild_id),
            )
            balance = await self._total_locked(conn, user_id, guild_id)
            await self._journal(conn, user_id, guild_id, kind, -amount, balance, note)
        return await self.get_account(user_id, guild_id)

    async def transfer(self, user_id: int, guild_id: int, amount: int) -> Account:
        """Move credits between wallet and bank in a single transaction."""
        account = await self.get_account(user_id, guild_id)
        if amount <= 0:
            raise ValueError("transfer amount must be positive")
        if account.wallet < amount:
            raise InsufficientFunds(amount, account.wallet)
        if account.bank + amount > account.bank_capacity:
            raise BankFull(account.bank_capacity)
        async with self.db.transaction() as conn:
            await conn.execute(
                """
                UPDATE accounts
                SET wallet = wallet - ?, bank = bank + ?, updated_at = unixepoch('subsec')
                WHERE user_id = ? AND guild_id = ?
                """,
                (amount, amount, user_id, guild_id),
            )
            await self._journal(
                conn, user_id, guild_id, "deposit", 0, account.total, f"Deposited {amount:,}"
            )
        return await self.get_account(user_id, guild_id)

    async def withdraw(self, user_id: int, guild_id: int, amount: int) -> Account:
        account = await self.get_account(user_id, guild_id)
        if amount <= 0:
            raise ValueError("withdrawal amount must be positive")
        if account.bank < amount:
            raise InsufficientFunds(amount, account.bank)
        async with self.db.transaction() as conn:
            await conn.execute(
                """
                UPDATE accounts
                SET wallet = wallet + ?, bank = bank - ?, updated_at = unixepoch('subsec')
                WHERE user_id = ? AND guild_id = ?
                """,
                (amount, amount, user_id, guild_id),
            )
            await self._journal(
                conn, user_id, guild_id, "withdraw", 0, account.total, f"Withdrew {amount:,}"
            )
        return await self.get_account(user_id, guild_id)

    async def upgrade_bank(self, user_id: int, guild_id: int) -> tuple[Account, int]:
        """Purchase the next bank tier. Returns the account and the cost paid."""
        account = await self.get_account(user_id, guild_id)
        cost = config.economy.bank_upgrade_base_cost * account.bank_level
        if account.wallet < cost:
            raise InsufficientFunds(cost, account.wallet)
        async with self.db.transaction() as conn:
            await conn.execute(
                """
                UPDATE accounts
                SET wallet = wallet - ?,
                    bank_level = bank_level + 1,
                    bank_capacity = bank_capacity + ?,
                    updated_at = unixepoch('subsec')
                WHERE user_id = ? AND guild_id = ?
                """,
                (cost, config.economy.bank_upgrade_capacity, user_id, guild_id),
            )
            balance = await self._total_locked(conn, user_id, guild_id)
            await self._journal(
                conn,
                user_id,
                guild_id,
                "bank_upgrade",
                -cost,
                balance,
                f"Bank upgraded to level {account.bank_level + 1}",
            )
        return await self.get_account(user_id, guild_id), cost

    # ------------------------------------------------------------- cooldowns

    @staticmethod
    def _remaining(last: float | None, cooldown: int) -> float:
        if last is None:
            return 0.0
        return max(0.0, (last + cooldown) - time.time())

    async def work(self, user_id: int, guild_id: int) -> tuple[int, Account]:
        account = await self.get_account(user_id, guild_id)
        remaining = self._remaining(account.last_work, config.economy.work_cooldown_seconds)
        if remaining > 0:
            raise CooldownActive(remaining)
        payout = random.randint(config.economy.work_min, config.economy.work_max)
        await self.db.execute(
            "UPDATE accounts SET last_work = ? WHERE user_id = ? AND guild_id = ?",
            (time.time(), user_id, guild_id),
        )
        updated = await self.credit(
            user_id, guild_id, payout, kind="work", note="Shift wages"
        )
        return payout, updated

    async def daily(self, user_id: int, guild_id: int) -> tuple[int, int, Account]:
        """Claim the daily reward. Returns (payout, streak, account)."""
        account = await self.get_account(user_id, guild_id)
        remaining = self._remaining(account.last_daily, config.economy.daily_cooldown_seconds)
        if remaining > 0:
            raise CooldownActive(remaining)

        # The streak survives one missed day; beyond that it resets.
        now = time.time()
        if account.last_daily is None or now - account.last_daily > config.economy.daily_cooldown_seconds * 2:
            streak = 1
        else:
            streak = min(account.daily_streak + 1, config.economy.daily_streak_cap)

        payout = config.economy.daily_amount + config.economy.daily_streak_bonus * (streak - 1)
        async with self.db.transaction() as conn:
            await conn.execute(
                """
                UPDATE accounts
                SET last_daily = ?, daily_streak = ?, updated_at = unixepoch('subsec')
                WHERE user_id = ? AND guild_id = ?
                """,
                (now, streak, user_id, guild_id),
            )
            await conn.execute(
                """
                UPDATE accounts SET bank = MIN(bank + ?, bank_capacity)
                WHERE user_id = ? AND guild_id = ?
                """,
                (payout, user_id, guild_id),
            )
            balance = await self._total_locked(conn, user_id, guild_id)
            await self._journal(
                conn,
                user_id,
                guild_id,
                "daily",
                payout,
                balance,
                f"Daily reward (streak {streak})",
            )
        return payout, streak, await self.get_account(user_id, guild_id)

    async def rob(
        self, user_id: int, guild_id: int, target_id: int
    ) -> tuple[bool, int, Account]:
        """Attempt a robbery. Returns (succeeded, amount, robber account)."""
        if user_id == target_id:
            raise ValueError("you cannot rob yourself")
        robber = await self.get_account(user_id, guild_id)
        remaining = self._remaining(robber.last_rob, config.economy.rob_cooldown_seconds)
        if remaining > 0:
            raise CooldownActive(remaining)
        victim = await self.get_account(target_id, guild_id)

        await self.db.execute(
            "UPDATE accounts SET last_rob = ? WHERE user_id = ? AND guild_id = ?",
            (time.time(), user_id, guild_id),
        )

        if victim.wallet < 100 or random.random() > config.economy.rob_success_chance:
            fine = min(config.economy.rob_fine, robber.wallet)
            if fine > 0:
                await self.debit(user_id, guild_id, fine, kind="rob_fine", note="Failed robbery")
            return False, fine, await self.get_account(user_id, guild_id)

        stolen = int(victim.wallet * random.uniform(0.05, config.economy.rob_max_steal_fraction))
        stolen = max(1, min(stolen, victim.wallet))
        async with self.db.transaction() as conn:
            await conn.execute(
                "UPDATE accounts SET wallet = wallet - ? WHERE user_id = ? AND guild_id = ?",
                (stolen, target_id, guild_id),
            )
            await conn.execute(
                "UPDATE accounts SET wallet = wallet + ? WHERE user_id = ? AND guild_id = ?",
                (stolen, user_id, guild_id),
            )
            victim_balance = await self._total_locked(conn, target_id, guild_id)
            await self._journal(
                conn, target_id, guild_id, "robbed", -stolen, victim_balance, "Robbery"
            )
            robber_balance = await self._total_locked(conn, user_id, guild_id)
            await self._journal(
                conn, user_id, guild_id, "robbery", stolen, robber_balance, "Successful robbery"
            )
        return True, stolen, await self.get_account(user_id, guild_id)

    # ------------------------------------------------------------- internals

    @staticmethod
    async def _total_locked(conn, user_id: int, guild_id: int) -> int:
        async with conn.execute(
            "SELECT wallet + bank FROM accounts WHERE user_id = ? AND guild_id = ?",
            (user_id, guild_id),
        ) as cursor:
            row = await cursor.fetchone()
            return int(row[0]) if row else 0

    @staticmethod
    async def _journal(
        conn,
        user_id: int,
        guild_id: int,
        kind: str,
        amount: int,
        balance_after: int,
        note: str | None,
    ) -> None:
        await conn.execute(
            """
            INSERT INTO transactions (user_id, guild_id, kind, amount, balance_after, note)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (user_id, guild_id, kind, amount, balance_after, note),
        )

    @staticmethod
    def _row_to_account(row) -> Account:
        return Account(
            user_id=row["user_id"],
            guild_id=row["guild_id"],
            wallet=int(row["wallet"]),
            bank=int(row["bank"]),
            bank_capacity=int(row["bank_capacity"]),
            bank_level=int(row["bank_level"]),
            daily_streak=int(row["daily_streak"]),
            last_daily=row["last_daily"],
            last_work=row["last_work"],
            last_rob=row["last_rob"],
        )
